### ISSUE TITLE
AlphaGomoku (GPU) making stupid moves

### ISSUE THREAD

**Issue by felix-22** [MaciejKozarzewski/AlphaGomoku#1]
The CPU version is not affected.
Also I noticed that GPU depths are low compared to CPU, despite having much more speed (n/s):
`GPU:  4) pbrain-AlphaGomoku.exe: depth 1-6 ev 50.0 n 31186 n/s 7136 tm 4370`
`CPU: 4) pbrain-AlphaGomoku.exe: depth 1-19 ev 48.4 n 4548 n/s 1035 tm 4390`
Any idea what's going on? 

![AG_stupid_moves](https://user-images.githubusercontent.com/75935068/129591594-751204da-fdb0-4903-baa3-a93db61105c0.gif)

**Comment by nguyencongminh090** [comment]
Maybe you run it in a too short time (time match = 5 seconds).

**Comment by felix-22** [comment]
Nope, it's 5s/120s move/match, furthermore even before reporting the issue, I already tried with 30s/1000s as well and it made no difference.
![settings](https://user-images.githubusercontent.com/75935068/129849762-388dfcb4-467a-4d57-a7ad-6155824fdeeb.gif)

**Comment by nguyencongminh090** [comment]
Did you try at 300s/300s turn/match?

**Comment by felix-22** [comment]
Yep, but it did not fix the issue neither.

**Comment by nguyencongminh090** [comment]
Did you mean in 300s/300s turn/match still making a bad move? Okay, wait for Maciej Kozarzewski response.

**Comment by nguyencongminh090** [comment]
Can I see your config file?

**Comment by felix-22** [comment]
The problem with logging solved.

**Comment by MaciejKozarzewski** [comment]
Oh, I didn't see your issue earlier!
Can you paste here a log file where this problem happened and a config file you use?

**Comment by felix-22** [comment]
[logs.zip](https://github.com/MaciejKozarzewski/AlphaGomoku/files/7022384/logs.zip)

**Comment by MaciejKozarzewski** [comment]
I see what's causing the problem. The neural network is not running at all (it returns zeros everywhere). I don't know why the engine didn't report that the network is not working, this is certainly a bug to be solved.

I guess that your GPU is no longer supported, either by NVIDIA math libraries or my own CUDA code. Which CUDA runtime version do you have? The last one supporting GPUs with compute capability 3.0 was

> CUDA SDK 10.0 – 10.2 support for compute capability 3.0 – 7.5 (Kepler, Maxwell, Pascal, Volta, Turing). Last version with support for compute capability 3.x (Kepler). 10.2 is the last official release for macOS, as support will not be available for macOS in newer releases.

I'm afraid I will not be able to workaround the NVIDIA policy of dropping support for older cards. The only solution I can recommend is to install older CUDA version.

**Comment by MaciejKozarzewski** [comment]
Now I think that installing older CUDA might not solve the problem. I would have to compile the code to include compute capability 3.0 using the older version of CUDA SDK than the one I currently use. I can do it, but in the future to support tensor cores on newest RTX 3xxx cards I would have to switch to newer CUDA, and drop the support for older cards anyway.

I'm afraid I can't solve this problem right now but I plan to include OpenCL support in the future versions. The older cards could be then run as OpenCL device and be free from such compatibility issues.

**Comment by felix-22** [comment]
> MaciejKozarzewski : https://github.com/MaciejKozarzewski/AlphaGomoku/issues/1#issuecomment-902806800 : Which CUDA runtime version do you have?

10.0, but I've tried with 9 right now and AG doesn't work either.
Katagomo works fine on CUDA 10.0 (and OpenCl as well).

**Comment by MaciejKozarzewski** [comment]
As I wrote, this is a problem on my side. In the next version of AG I can compile with CUDA 10.0 to include compute capability 3.0 (Katagomo must have been compiled in such a way to work) as I don't make use of tensor cores at the moment. But in the future, when I add OpenCL support, I'll revert to using newer CUDA versions.
